embassy-stm32/src/flash/mod.rs

```diff
@@ -17,74 +19,175 @@ pub struct Flash<'d> {
 
         unsafe {
+            family::clear_all_err();
             family::unlock();
-            let res = family::blocking_erase(from, to);
+            let res = family::blocking_erase(start_address, end_address);
             family::lock();
             res
         }
     }
+}
 
-    fn clear_all_err(&mut self) {
-        unsafe { family::clear_all_err() };
+impl Drop for Flash<'_> {
+    fn drop(&mut self) {
+        unsafe { family::lock() };
     }
 }
 
-impl Drop for Flash<'_> {
+impl Drop for FlashRegions {
     fn drop(&mut self) {
         unsafe { family::lock() };
     }
 }
 
+pub trait FlashRegion {
+    const BASE: usize;
+    const SIZE: usize;
+    const ERASE_SIZE: usize;
+    const WRITE_SIZE: usize;
+    const ERASE_VALUE: u8;
+
+    fn blocking_read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Error> {
+        if offset as usize + bytes.len() > Self::SIZE {
+            return Err(Error::Size);
+        }
+
+        let first_address = Self::BASE as u32 + offset;
+        let flash_data = unsafe { core::slice::from_raw_parts(first_address as *const u8, bytes.len()) };
+        bytes.copy_from_slice(flash_data);
+        Ok(())
+    }
+
+    fn blocking_write(&mut self, offset: u32, buf: &[u8]) -> Result<(), Error> {
+        if offset as usize + buf.len() > Self::SIZE {
+            return Err(Error::Size);
+        }
+        if offset as usize % Self::WRITE_SIZE != 0 || buf.len() as usize % Self::WRITE_SIZE != 0 {
+            return Err(Error::Unaligned);
+        }
+
+        let start_address = Self::BASE as u32 + offset;
+        trace!("Writing {} bytes from 0x{:x}", buf.len(), start_address);
+
+        // Protect agains simultaneous write/erase to multiple regions.
+        let _guard = take_lock_spin();
+
+        unsafe {
+            family::clear_all_err();
+            family::unlock();
+            let res = Flash::blocking_write_all(start_address, buf);
+            family::lock();
+            res
+        }
+    }
+
+    fn blocking_erase(&mut self, from: u32, to: u32) -> Result<(), Error> {
+        if to < from || to as usize > Self::SIZE {
+            return Err(Error::Size);
+        }
+        if (from as usize % Self::ERASE_SIZE) != 0 || (to as usize % Self::ERASE_SIZE) != 0 {
+            return Err(Error::Unaligned);
+        }
+
+        let start_address = Self::BASE as u32 + from;
+        let end_address = Self::BASE as u32 + to;
+        trace!("Erasing from 0x{:x} to 0x{:x}", start_address, end_address);
+
+        // Protect agains simultaneous write/erase to multiple regions.
+        let _guard = take_lock_spin();
+
+        unsafe {
+            family::clear_all_err();
+            family::unlock();
+            let res = family::blocking_erase(start_address, end_address);
+            family::lock();
+            res
+        }
+    }
+}
+
+fn take_lock_spin() -> MutexGuard<'static, CriticalSectionRawMutex, ()> {
```

Comment:
never use spinlocks. They can cause deadlocks if you spin on a higher priority context waiting for a lower priority context to release it.

In this particular case, it should be enough to wrap all the operations that could race with `critical_section::with`.

Reply:
As stated above, I could use a suggestion on how to do this without `critical_section::with` as I need some way of:
1. unlock()
2. write()/erase()
3. lock()

where I am guaranteed that I do the entire sequence. BUT, I do not want to disable interrupt during the entire sequence. The reason is that it may take a very long time to write/erase. I know that the mcu may stall during e.g. erase, but only per sector. It will wake up and handle higher priority interrupts between each sector. Wrapping the entire sequence with a cs will multiply the time with the sector count which I think we can do better.

Reply:
Agreed, and it should be doable, yes! probably just taking a CS fo each operation? 

Also perhaps lock/unlock for each operation. It feels dangerous to let irqs fire while flash is unlocked (though the current code already does that, so idk). lock/unlock is instant compared to erase/write, so it shouldn't make a difference in perf.

```rust
for sector in some_sector_iter {
   critical_section::with(|_| {
      unlock();
      erase(sector);
      lock();
   });
}
```

Reply:
Done!